src/packageManager/downloadAndInstallPackages.ts

```diff
@@ -16,14 +16,39 @@ import { mkdirpSync } from 'fs-extra';
 import { PackageInstallStart } from '../shared/loggingEvents';
 import { DownloadValidator } from './isValidDownload';
 import { CancellationToken } from 'vscode';
+import { ITelemetryReporter } from '../shared/telemetryReporter';
 
 export async function downloadAndInstallPackages(
     packages: AbsolutePathPackage[],
     provider: NetworkSettingsProvider,
     eventStream: EventStream,
     downloadValidator: DownloadValidator,
+    telemetryReporter?: ITelemetryReporter,
     token?: CancellationToken
 ): Promise<boolean> {
+    function sendInstallationFailureTelemetry(pkg: AbsolutePathPackage, installationStage: string, error: any): void {
+        if (!telemetryReporter) {
+            return;
+        }
+
+        const telemetryProperties: { [key: string]: string } = {
+            installStage: installationStage,
+            packageId: pkg.id,
+            isOptional: pkg.isOptional ? 'true' : 'false',
+        };
+
+        if (error instanceof NestedError && error.err instanceof PackageError) {
+            telemetryProperties['error.message'] = error.err.message;
+            telemetryProperties['error.packageUrl'] = error.err.pkg.url;
+        } else if (error instanceof PackageError) {
+            telemetryProperties['error.message'] = error.message;
+            telemetryProperties['error.packageUrl'] = error.pkg.url;
+        }
+
+        const eventName = pkg.isOptional ? 'OptionalPackageInstallationFailed' : 'PackageInstallationFailed';
```

Comment:
think its probably not necessary to have a different event name here, we already set a property on the event indicating its optional or not.